### PR TITLE
ci: use GitHub app for ephemeral tokens

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Use the GitHub app to generate the required ephemeral tokens with the least permissive principle.

#### Why

* Finer-grained tokens with Service Machine accounts are required to rotate the secrets manually.
* GitHub app to generate temporary tokens is the advanced approach to avoid the above
* Document what the GH workflow requires to run in terms of access
* GitHub Token with Permissions does not trigger GitHub builds 

#### Implementation details

Use `tibdex/github-app-token` with the required permissions and the repository scope
Remove the `permissions` configuration in the GH workflow.
Configure git checkout with the ephemeral token
Configure the GH_TOKEN with the ephemeral token

## Checklist

- [ ] This is an enhancement of existing features, or a new feature in existing plugins
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation
- [ ] This is a bugfix
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [ ] This is something else
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)

### CODE REVIEW

**Review by reakaleek — APPROVED** [review]

File: .github/workflows/release-step-3.yml [hunk 1]
```diff
@@ -213,17 +213,28 @@ jobs:
...
     steps:
+      - name: Get token
+        id: get_token
+        uses: tibdex/github-app-token@3beb63f4bd073e61482598c45c71c1019b59b73a # v2.1.0
+        with:
+          app_id: ${{ secrets.OBS_AUTOMATION_APP_ID }}
+          private_key: ${{ secrets.OBS_AUTOMATION_APP_PEM }}
+          permissions: >-
+            {
+              "contents": "write"
```
Comment (v1v) [inline .github/workflows/release-step-3.yml:225]: IIUC, this step is only for updating a GitHub branch, hence the permission has been changed

Comment (v1v) [reply .github/workflows/release-step-3.yml:225]: wrong assumption:
- https://github.com/elastic/apm-agent-java/blob/9a12b98ee959abb4a596c823325d9c78d827724b/.ci/release/update-major-branch.sh#L70-L75
It creates a PR too

Comment (v1v) [reply .github/workflows/release-step-3.yml:225]:
```suggestion
              "contents": "write",
              "pull_requests": "write"
```

File: .github/workflows/pre-post-release.yml [hunk 2]
```diff
@@ -53,17 +53,29 @@ jobs:
...
+        with:
+          app_id: ${{ secrets.OBS_AUTOMATION_APP_ID }}
+          private_key: ${{ secrets.OBS_AUTOMATION_APP_PEM }}
+          permissions: >-
+            {
+              "contents": "write",
+              "pull_requests": "write"
+            }
+          repositories: >-
+            ["${{github.repository}}"]
```
Comment (v1v) [inline .github/workflows/pre-post-release.yml:69]: See https://github.com/tibdex/github-app-token/issues/104 or https://github.com/tibdex/github-app-token/issues/106
```suggestion
            ["apm-agent-java"]
```

File: .github/workflows/release-step-3.yml [hunk 3]
```diff
@@ -213,17 +213,29 @@ jobs:
...
+        with:
+          app_id: ${{ secrets.OBS_AUTOMATION_APP_ID }}
+          private_key: ${{ secrets.OBS_AUTOMATION_APP_PEM }}
+          permissions: >-
+            {
+              "contents": "write",
+              "pull_requests": "write"
+            }
+          repositories: >-
+            ["${{github.repository}}"]
```
Comment (v1v) [inline .github/workflows/release-step-3.yml:229]: See https://github.com/tibdex/github-app-token/issues/104 or https://github.com/tibdex/github-app-token/issues/106
```suggestion
            ["apm-agent-java"]
```

**Review by SylvainJuge — APPROVED** [review]
LGTM, do we have a way to test this before next release ?

**Comment by v1v** [comment]
> LGTM, do we have a way to test this before next release ?

I tested the GH app was doing what's expected, generating the tokens in other different places. I cannot think of a way to test this in isolation in dry-run mode since it will not utilise the token.

I guess the `pre-release` step should help to know if things are ok - can we run that step after merging this PR to confirm things work as expected?

**Pull request merged**